modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -1994,4 +2070,252 @@ private void processDelay(InetSocketAddress sockAddr) {
+                    throw new BinaryObjectException(e);
+                }
+            }
+
+            return meta;
+        }
+
+        /**
+         * Handle update binary schema result.
+         *
+         * @param res Result.
+         * @return {@code true} if handler was waiting for result with given request ID.
+         */
+        public boolean handleResult(JdbcUpdateBinarySchemaResult res) {
+            return handleResult(res.reqId(), res);
+        }
+
+        /**
+         * Handle binary type schema result.
+         *
+         * @param res Result.
+         * @return {@code true} if handler was waiting for result with given request ID.
+         */
+        public boolean handleResult(JdbcBinaryTypeGetResult res) {
+            return handleResult(res.reqId(), res);
+        }
+
+        /** {@inheritDoc} */
+        @Override public BinaryType metadata(int typeId, int schemaId) throws BinaryObjectException {
+            BinaryType type = metadata(typeId);
+            return type != null && ((BinaryTypeImpl)type).metadata().hasSchema(schemaId) ? type : null;
+        }
+
+        /** {@inheritDoc} */
+        @Override public Collection<BinaryType> metadata() throws BinaryObjectException {
+            return cache.metadata();
+        }
+    }
+
+    /**
+     * Jdbc channel to communicate in blocking style, regardless of whether streaming mode is enabled or not.
+     */
+    private abstract class BlockingJdbcChannel {
+        /** Request ID -> Jdbc result map. */
+        private Map<Long, CompletableFuture<JdbcResult>> results = new ConcurrentHashMap<>();
+
+        /**
+         * Handles result for specified request ID.
+         *
+         * @param reqId Request id.
+         * @param res Result.
+         */
+        boolean handleResult(long reqId, JdbcResult res) {
+            boolean handled = false;
+            CompletableFuture<JdbcResult> fut = results.remove(reqId);
+            if (fut != null) {
+                fut.complete(res);
+                handled = true;
+            }
+            return handled;
+        }
+
+        /**
+         * Do request in blocking style. It just call {@link JdbcThinConnection#sendRequest(JdbcRequest)} for
+         * non-streaming mode and creates future and waits it completion when streaming is enabled.
+         *
+         * @param req Request.
+         * @return Result for given request.
+         */
+        <R extends JdbcResult> R doRequest(
+            JdbcRequest req) throws SQLException, InterruptedException, ExecutionException {
+            R res;
+            if (isStream()) {
+                CompletableFuture<JdbcResult> resFut = new CompletableFuture<>();
+
+                assert results.put(req.requestId(), resFut) == null : "Another request with the same id is waiting for result.";
+
+                sendRequestNotWaitResponse(req, streamState.streamingStickyIo);
```

Comment:
Why cannot we use `sendRequest(req)` if streaming is on?

Reply:
because we'll have race for input stream of responses.